inseq/attr/feat/ops/basic_attention.py

```diff
@@ -0,0 +1,247 @@
+
+        Returns:
+            `torch.Tensor`: The attention tensor with its attention heads merged.
+        """
+        num_heads = self._num_attention_heads(attention[0])
+
+        if option == "single" and head is None:
+            raise RuntimeError("An attention head has to be specified when choosing single-head attention attribution")
+
+        if head is not None:
+            if head > num_heads:
+                raise RuntimeError(
+                    f"Attention head index for attribution too high. The model only has {num_heads} heads."
+                )
+
+            if option != "single":
+                logger.warning(
+                    "Only single-head attention is possible if an attention head is specified.\n"
+                    "Switching to single-head attention"
+                )
+
+            return attention.select(1, head)
+
+        if option == "average":
+            return attention.mean(1)
+
+        elif option == "max":
+            return attention.max(1)
+
+        else:
+            raise RuntimeError(
+                "Invalid merge method for attention heads specified. "
+                "Valid methods are: `average`, `max` and `single`"
+            )
+
+    def _get_batch_size(self, attention: torch.Tensor) -> int:
+        """returns the batch size of a tensor of shape `(batch_size, heads, seq, seq)`"""
+        return attention.size(0)
+
+    def _extract_forward_pass_args(
+        self,
+        inputs: MultiStepEmbeddingsTensor,
+        forward_args: Optional[Tuple],
+        is_target_attr: bool,
+        is_encoder_decoder: bool,
+    ) -> dict:
+        """extracts the arguments needed for a standard forward pass
+        from the `inputs` and `additional_forward_args` parameters used by Captum"""
+
+        if is_encoder_decoder:
+
+            use_embeddings = forward_args[6] if is_target_attr else forward_args[7]
+
+            forward_pass_args = {
+                "attention_mask": forward_args[4] if is_target_attr else forward_args[5],
+                "decoder_attention_mask": forward_args[5] if is_target_attr else forward_args[6],
+            }
+
+            if use_embeddings:
+                forward_pass_args["inputs_embeds"] = inputs[0]
+                forward_pass_args["decoder_inputs_embeds"] = inputs[1] if is_target_attr else forward_args[0]
+            else:
+                forward_pass_args["input_ids"] = forward_args[0] if is_target_attr else forward_args[1]
+                forward_pass_args["decoder_input_ids"] = forward_args[1] if is_target_attr else forward_args[2]
+
+        else:
+
+            use_embeddings = forward_args[4]
+
+            forward_pass_args = {"attention_mask": forward_args[3]}
+
+            if use_embeddings:
+                forward_pass_args["inputs_embeds"] = inputs[0]
+            else:
+                forward_pass_args["input_ids"] = forward_args[0]
+
+        return forward_pass_args
```

Comment:
This function is superfluous. The method used for extracting step scores can be recycled here, and does not require explicit knowledge of model type inside here, since the correct method will be called from the model class (see [here](https://github.com/inseq-team/inseq/blob/9e88a9ebf3788043ec6cc0c6fa334f594a29b154/inseq/attr/feat/attribution_utils.py#L107) how it is used)